src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1434 @@
+              block->size, coarse_provider->used_size - block->size,
+              coarse_provider->alloc_size);
+
+    assert(coarse_provider->used_size >= block->size);
+    coarse_provider->used_size -= block->size;
+
+    block->used = false;
+
+    // Merge with prev and/or next block if they are unused and have continuous data.
+    block = free_block_merge_with_prev(
+        coarse_provider->upstream_memory_provider, coarse_provider->all_blocks,
+        coarse_provider->free_blocks, block);
+    block = free_block_merge_with_next(
+        coarse_provider->upstream_memory_provider, coarse_provider->all_blocks,
+        coarse_provider->free_blocks, block);
+
+    int rv = free_blocks_add(coarse_provider->free_blocks, block);
+    if (rv) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    assert(debug_check(coarse_provider));
+
+    if (utils_mutex_unlock(&coarse_provider->lock) != 0) {
+        LOG_ERR("unlocking the lock failed");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void coarse_memory_provider_get_last_native_error(void *provider,
+                                                         const char **ppMessage,
+                                                         int32_t *pError) {
+    (void)provider; // unused
+
+    if (ppMessage == NULL || pError == NULL) {
+        assert(0);
+        return;
+    }
+
+    // Nothing more is needed here, since
+    // there is no UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC error used.
+}
+
+static umf_result_t coarse_memory_provider_get_min_page_size(void *provider,
+                                                             void *ptr,
+                                                             size_t *pageSize) {
+    if (provider == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    if (!coarse_provider->upstream_memory_provider) {
+        *pageSize = utils_get_page_size();
+        return UMF_RESULT_SUCCESS;
+    }
+
+    return umfMemoryProviderGetMinPageSize(
+        coarse_provider->upstream_memory_provider, ptr, pageSize);
+}
+
+static umf_result_t
+coarse_memory_provider_get_recommended_page_size(void *provider, size_t size,
+                                                 size_t *pageSize) {
+    if (provider == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    if (!coarse_provider->upstream_memory_provider) {
+        *pageSize = utils_get_page_size();
+        return UMF_RESULT_SUCCESS;
+    }
+
+    return umfMemoryProviderGetRecommendedPageSize(
+        coarse_provider->upstream_memory_provider, size, pageSize);
+}
+
+static const char *coarse_memory_provider_get_name(void *provider) {
+    (void)provider;
+
+    return "coarse";
```

Comment:
Shouldn't we return name of upstream provider with coarse prefix/sufix. 

"coarse (L0)" or "Level Zero (coarse)"

Reply:
@bratpiorka what do you think about it ?

Reply:
I agree with the combination of "coarse" and upstream->get_name(). I prefer "coarse (L0)" option

Reply:
@lplewa Done